### PR TITLE
Fix bugs in DeltaFIFO

### LINKED ISSUE #27004: Race condition in Informer
### What's the issue?
Informer could end up in wrong state.

### Example?
I have been stress testing RC of 100 replicas. I kept deleting the pod and let RC recover it to 100. The program will do informer.store.List() to check if there are 100 pods. But at the end of day it returned more than 100 (e.g. 131), while listing directly from apiserver returned 100.

### Why?
Informer has a queue and a store (aka indexer) internally.
- The store relies on the queue to do blocking [pop()][1] and and then [updates itself][2] in [Process()][1].
- The queue (delta fifo) is used to queue events from watcher. For [each operation in reflector][3], e.g. Delete, we will [enqueue an event to the queue][4].
- However, in DeltaFIFO.Delete(), it will [check the store][5] first to see if it needs to send a delete event. **This logic is wrong because the queue and store are run concurrently. There could be an ADD event not processed by store yet thus queue won't enqueue DELETE event.**

[1]:https://github.com/kubernetes/kubernetes/blob/0d3be6a316b39a71b62d45272d339f7fa9162165/pkg/controller/framework/controller.go#L127-L128
[2]:https://github.com/kubernetes/kubernetes/blob/0d3be6a316b39a71b62d45272d339f7fa9162165/pkg/controller/framework/shared_informer.go#L200-L224
[3]:https://github.com/kubernetes/kubernetes/blob/0d3be6a316b39a71b62d45272d339f7fa9162165/pkg/client/cache/reflector.go#L384-L392
[4]:https://github.com/kubernetes/kubernetes/blob/0d3be6a316b39a71b62d45272d339f7fa9162165/pkg/client/cache/delta_fifo.go#L200
[5]:https://github.com/kubernetes/kubernetes/blob/0d3be6a316b39a71b62d45272d339f7fa9162165/pkg/client/cache/delta_fifo.go#L192-L198

### PR DESCRIPTION
**Pull request by wojtek-t**
Fix #27004

### CODE REVIEW

**Comment by wojtek-t** [comment]
@hongchaodeng @caesarxuchao @lavalamp

File: pkg/client/cache/fifo.go [hunk 1]
```diff
@@ -45,6 +45,8 @@ type Queue interface {
 }
 
 // Helper function for popping from Queue.
+// WARNING: Do NOT use this function in non-test code to avoid races
+// unless you really really really really know what you are doing.
```
Comment (dims) [inline pkg/client/cache/fifo.go]: 4 really's are really not enough :)

File: pkg/client/cache/delta_fifo.go [hunk 2]
```diff
@@ -189,12 +189,20 @@ func (f *DeltaFIFO) Delete(obj interface{}) error {
...
+		// We only want to skip the "deletion" action if the object doesn't
+		// exist in knownObjects and it doesn't have corresponding item in items.
+		// Note that even if there is a "deletion" action in items, we can ignore it,
+		// because it will be deduped automatically in "queueActionLocked"
+		_, exists, err := f.knownObjects.GetByKey(id)
+		_, itemsExist := f.items[id]
+		if err == nil && !exists && !itemsExist {
+			// Presumably, this was deleted when a relist happened.
+			// Don't provide a second report of the same deletion.
+			// TODO(lavalamp): This may be racy-- we aren't properly locked
```
Comment (lavalamp) [inline pkg/client/cache/delta_fifo.go]: This TODO is no longer true, right?

Comment (wojtek-t) [inline pkg/client/cache/delta_fifo.go]: I think so.

**Comment by lavalamp** [comment]
LGTM, thanks a bunch!

File: pkg/client/cache/delta_fifo.go [hunk 3]
```diff
@@ -189,12 +189,20 @@ func (f *DeltaFIFO) Delete(obj interface{}) error {
...
-	} else if _, exists, err := f.knownObjects.GetByKey(id); err == nil && !exists {
-		// Presumably, this was deleted when a relist happened.
-		// Don't provide a second report of the same deletion.
-		// TODO(lavalamp): This may be racy-- we aren't properly locked
-		// with knownObjects.
-		return nil
+	} else {
+		// We only want to skip the "deletion" action if the object doesn't
+		// exist in knownObjects and it doesn't have corresponding item in items.
+		// Note that even if there is a "deletion" action in items, we can ignore it,
```
Comment (caesarxuchao) [inline pkg/client/cache/delta_fifo.go]: This is clever. Thanks a bunch.

**Comment by lavalamp** [comment]
Manual build cop merge.

**Pull request merged**